### DISCUSSION TITLE
Transitive dependencies of components

### DISCUSSION THREAD

**Discussion by billcrook** [DavidVujic/python-polylith#85] [General]
Thanks for the interesting project. I've just started to dig in but one thing that stood out is the need to redeclare transitive dependencies of components. For instance:
* component A has a dependency on, for instance, pendulum
* project X has a dependency on component A
* project X has to declare the dependency on pendulum
* project Y would have to also declare if it depends on component A
* top-level project also has to declare dependency on pendulum, presumably for testing

So, the duplication of pendulum dependency can end up being N+1 where N=number of projects that use component A.

Ideally one would declare the pendulum dependency _within_ component A once.

Any thoughts on this?

**Comment by DavidVujic** [comment]
Hi @billcrook, thank you for sharing feedback!

Yes, that's how it is set up today. Each project need to have dependencies added in their respective `pyproject.toml`. The Clojure implementation of Polylith has a nice approach to this and I think it works just as you suggests. There's an option to add a `deps.edn` in a component, and that will be picked up by each project using it.

I have that as a long-term goal for Python too, but believe there's a couple of things that need some experimentation and trying out. I have chosen to implement Polylith components and bases (aka bricks) as Python namespace packages. Initially I had an idea to have them as editable dependencies (libraries), to define the dependencies for a brick in there. But I leaned towards namespace packages instead, to avoid versioning, conflicting dependencies (such as different versions of a library) between bricks. 

As now, the tooling support will analyze code and highlight any missing dependencies (the `poly libs` and `poly check` commands) and am currently working on adding a new `sync` command that will add missing code into a project automatically. Initially, it will be missing Polylith bricks only, but maybe also third-party dependencies.

Another thing that I have in the back of my head long-term is to implement something in-between to what's in Clojure and the third-party dependencies in Python: a Polylith-specific toml-file (or similar) for components. Maybe that's a bad idea, not sure 😄

**Comment by billcrook** [comment]
Great, glad you are thinking about it.  For larger repos with many components and projects, I imagine the duplication of dependencies could get quite cumbersome.

> Another thing that I have in the back of my head long-term is to implement something in-between to what's in Clojure and the third-party dependencies in Python: a Polylith-specific toml-file (or similar) for components. Maybe that's a bad idea, not sure

My two cents, I would go as far as I could with standard dependency tooling first. IDE compatibility is important and a custom toml file would be a non-starter for some people. If you could get it to work where each component has a poetry pyproject.toml, I think IDEs like pycharm would have a much easier time with it.

I have a few personal side-projects that would be better managed in this polylith style so I'm going to try it out. Thanks a bunch.
